### PR TITLE
8292983: ModuleReferenceImpl.computeHash should record algorithm for cache checks

### PR DESCRIPTION
**Pull request by shipilev**
Look at implementation and figure out what happens if you do:

```
 computeHash("SHA-1") = someHash;
 computeHash("SHA-256") = ...?
```

The caching method should actually check the algorithms match.

Not a bug at this point, since only use SHA-256 today, but this is a landmine ready to fire. 

Additional testing:
 - [x] Linux x86_64 release, `java/lang/module` tests

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292983](https://bugs.openjdk.org/browse/JDK-8292983): ModuleReferenceImpl.computeHash should record algorithm for cache checks

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10044/head:pull/10044` \
`$ git checkout pull/10044`

Update a local copy of the PR: \
`$ git checkout pull/10044` \
`$ git pull https://git.openjdk.org/jdk pull/10044/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10044`

View PR using the GUI difftool: \
`$ git pr show -t 10044`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10044.diff">https://git.openjdk.org/jdk/pull/10044.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java [hunk 1]
```diff
@@ -62,8 +62,18 @@ public class ModuleReferenceImpl extends ModuleReference {
...
+    // For safety reasons under concurrent updates, we need to wrap the
+    // hash/algo pair into the final wrapper.
+    private CachedHash cachedHash;
+    private final class CachedHash {
+        public CachedHash(byte[] hash, String algorithm) {
+            this.hash = hash;
+            this.algorithm = algorithm;
+        }
+        final byte[] hash;
+        final String algorithm;
```
Comment (AlanBateman) [inline src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java:79]: Have you tried make this a record? Also can you move the declaration of cachedHash to after the declaration of the record.

File: src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java [hunk 2]
```diff
@@ -139,13 +149,19 @@ public ModuleResolution moduleResolution() {
      * @throws java.io.UncheckedIOException if an I/O error occurs
      */
     public byte[] computeHash(String algorithm) {
-        byte[] result = cachedHash;
-        if (result != null)
-            return result;
-        if (hasher == null)
+        CachedHash ch = cachedHash;
+        if (ch != null) {
+            if (ch.algorithm.equals(algorithm)) {
```
Comment (AlanBateman) [inline src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java]: The nested if looks a bit strange here, you can change to:
```
if (ch != null & ch.algorithm.equals(algorithm))
    return ch.hash;
```

Comment (shipilev) [reply src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java]: Already done.

File: src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java [hunk 3]
```diff
@@ -62,8 +62,22 @@ public class ModuleReferenceImpl extends ModuleReference {
...
     private final ModuleResolution moduleResolution;
 
-    // cached hash of this module to avoid needing to compute it many times
-    private byte[] cachedHash;
+    // Single-slot cache of this module's hash to avoid needing to compute
+    // it many times. If there is a significant number of hash algorithms
+    // used in the future, consider extending this cache to full
+    // ConcurrentMap<String,byte[]>.
+    //
+    // For correctness under concurrent updates, we need to wrap the fields
```
Comment (AlanBateman) [inline src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java:70]: The comment "If there is a significant ..." reads more like a "TODO" and can be removed.

File: src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java (hunk 1 above)
Comment (shipilev) [reply src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java]: Switched to `record`, moved.

File: src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java (hunk 3 above)
Comment (shipilev) [reply src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java]: Done.

**Comment by AlanBateman** [comment]
Can you make sure that you run all jar, jmod and jlink tests? Module hashes are generated at packaging time, then checked when generating the configuration in early startup.

**Comment by shipilev** [comment]
> Can you make sure that you run all jar, jmod and jlink tests? Module hashes are generated at packaging time, then checked when generating the configuration in early startup.

`java/util/jar`, `jdk/tools/jmod`, `jdk/tools/jlink` -- all fine on Linux x86_64 release.

File: src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java [hunk 4]
```diff
@@ -62,8 +62,12 @@ public class ModuleReferenceImpl extends ModuleReference {
     // ModuleResolution flags
     private final ModuleResolution moduleResolution;
 
-    // cached hash of this module to avoid needing to compute it many times
-    private byte[] cachedHash;
+    // Single-slot cache of this module's hash to avoid needing to compute
+    // it many times. For correctness under concurrent updates, we need to
+    // wrap the fields updated at the same time with a record, which carries
+    // the implicit final-field semantics for its members.
+    private record CachedHash(byte[] hash, String algorithm) {}
```
Comment (AlanBateman) [inline src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java:69]: The comment still seems a bit excessive, maybe drop the last bit "which carries ..." as it is not needed here.

Comment (shipilev) [reply src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java:68]: I don't think the comment is excessive. It declares the intent to  protects from future "optimization" of flattening the record into the plain fields. But I dropped "which carries".

**Review by AlanBateman — APPROVED** [review]

**Review by mlchung — APPROVED** [review]
Looks good to me.

**Review by jaikiran — APPROVED** [review]

**Comment by shipilev** [comment]
Thank you all.

/integrate

**Pull request closed without merging**